### PR TITLE
vendor buildkit v0.10.0

### LINKED ISSUE #42261: (WSL2) docker build "output clipped, log limit 1MiB reached"
**Description**

When performing lengthy `docker build` operations using `DOCKER_BUILDKIT=1` (not using `buildx`, as it is experimental), the command stops outputting logs with the "_output clipped, log limit 1MiB reached_" message, until the build is over.

(This was first opened on https://github.com/docker/for-win/issues/10839, but I've been instructed to open it here instead)

**Steps to reproduce the issue:**
1. `DOCKER_BUILDKIT=1 docker build ....` that produces more than 1MiB of logs on stdout/stderr

**Describe the results you received:**
- Last output line says "_output clipped, log limit 1MiB reached_"
- No more log output in the terminal until the build is finished.

**Describe the results you expected:**
- Continuous log output in the terminal
- Or, clear instructions on how to increase that limit using stable functionality

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           20.10.3
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        48d30b5
 Built:             Fri Jan 29 14:33:13 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.5
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       363e9a8
  Built:            Tue Mar  2 20:15:47 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
  scan: Docker Scan (Docker Inc., v0.5.0)

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 7
 Server Version: 20.10.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.4.72-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 15.63GiB
 Name: docker-desktop
 ID: 3DGW:YZBD:OHOU:UUWW:J2JR:XV2Z:GCIF:3P43:6YAU:IJOX:7MAL:E2DI
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 49
  Goroutines: 52
  System Time: 2021-04-06T19:23:48.7791219Z
  EventsListeners: 3
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  localhost:5000
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

- Running on WSL2 (Hyper-V backend), on Windows 10 Enterprise, Version 20H2 (OS Build 19042.867)
- WSL2 VM OS: Ubuntu 18.04
- Can't use `docker buildx`, as it is an experimental feature, and I run in a production environment.

### PR DESCRIPTION
**Pull request by crazy-max**
closes #42968
closes #42983
closes #40023
fixes #42261

https://github.com/moby/buildkit/compare/9f254e1...8d45bd6

cc @thaJeztah @tonistiigi 

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

File: builder/builder-next/exporter/writer.go [hunk 1]
```diff
@@ -3,14 +3,15 @@ package containerimage
...
 
+	"github.com/containerd/containerd/platforms"
 	"github.com/moby/buildkit/cache"
+	binfotypes "github.com/moby/buildkit/util/buildinfo/types"
 	"github.com/moby/buildkit/util/progress"
 	"github.com/moby/buildkit/util/system"
-	digest "github.com/opencontainers/go-digest"
-	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
```
Comment (thaJeztah) [inline builder/builder-next/exporter/writer.go:14]: Probably good to keep this as `ocispec` (for now), as that's the alias we use elsewhere in the codebase (keeping it consistent makes it easier to find occurrences / uses of the same things). Also keep the `digest "github.com/opencontainers/go-digest"` (for now), I think there was some linter that preferred it, but if we remove it we should remove it everywhere (probably best to keep that separate from this PR)

File: builder/builder-next/worker/worker.go [hunk 2]
```diff
@@ -42,10 +40,9 @@ import (
 	"github.com/moby/buildkit/util/contentutil"
 	"github.com/moby/buildkit/util/progress"
 	digest "github.com/opencontainers/go-digest"
-	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
```
Comment (thaJeztah) [inline builder/builder-next/worker/worker.go:43]: Same here

File: Dockerfile.windows [hunk 3]
```diff
@@ -166,7 +166,7 @@ FROM microsoft/windowsservercore
 SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]
 
 ARG GO_VERSION=1.17.6
-ARG CONTAINERD_VERSION=1.5.9
+ARG CONTAINERD_VERSION=1.6.0-rc.2
```
Comment (thaJeztah) [inline Dockerfile.windows:169]: This is probably good for a separate PR (containerd binary should be compatible? if that's not the case, there's a problem)

**Review by thaJeztah — COMMENTED** [review]
some quick comments (my browser doesn't like reviewing the diff, so let me post some of them already; will have a closer look later)

File: builder/builder-next/exporter/export.go [hunk 4]
```diff
@@ -4,21 +4,35 @@ import (
...
+	"github.com/moby/buildkit/util/buildinfo"
+	"github.com/moby/buildkit/util/compression"
+	"github.com/opencontainers/go-digest"
 	"github.com/pkg/errors"
-	"github.com/sirupsen/logrus"
 )
 
 const (
-	keyImageName = "name"
+	keyImageName        = "name"
```
[the remark this answers is not in the document]
Comment (crazy-max) [reply builder/builder-next/exporter/export.go:26]: Yeah while working on this PR I thought about moving the moby exporter to https://github.com/moby/buildkit/tree/master/exporter. Maybe we could expose some bits on BuildKit in the meantime.

File: Dockerfile.windows (hunk 3 above)
Comment (crazy-max) [reply Dockerfile.windows:169]: Oh sure yes should be in a dedicated PR, was just testing its behavior with the goto setup.

File: builder/builder-next/controller.go [hunk 5]
```diff
@@ -268,18 +260,6 @@ func getGCPolicy(conf config.BuilderConfig, root string) ([]client.PruneInfo, er
 	return gcPolicy, nil
 }
 
-func parsePlatforms(platformsStr []string) ([]specs.Platform, error) {
```
[the remark this answers is not in the document]
Comment (crazy-max) [reply builder/builder-next/controller.go:271]: Yes BuildKit takes care of it now.

File: builder/builder-next/exporter/writer.go (hunk 1 above)
Comment (crazy-max) [reply builder/builder-next/exporter/writer.go:14]: Sure

**Comment by crazy-max** [comment]
Needs https://github.com/moby/moby/pull/41982

**Comment by tonistiigi** [comment]
This needs the additional patches for MergeOp/DiffOp from @sipsma . If we want to test sooner and reduce patch size we can first vendor the commit before MergeOp was merged.

**Comment by sipsma** [comment]
Commits needed to support changes made during mergeop:
* https://github.com/sipsma/moby/commit/b768ec3f6f7d70a18125f1213a7b917d74bcbbf6
* https://github.com/sipsma/moby/commit/e0759344b7f98956246bf744711ab7695b20e209

Same but for DiffOp:
* https://github.com/sipsma/moby/commit/3a5a2632104bb7610f97b24495d99dbba66e50d7

The changes needed to get the hardlink optimization working are here but are rough, need some cleanup:
* https://github.com/sipsma/moby/commit/e1dacf007cfc75f2ff44bd2035a898b4367a1a63

Let me know if I can help out cleaning the hardlink optimization commit up.

**Comment by crazy-max** [comment]
Thanks @sipsma!

File: builder/builder-next/exporter/export.go [hunk 6]
```diff
@@ -44,78 +46,109 @@ func New(opt Opt) (exporter.Exporter, error) {
...
-				i.targetNames = append(i.targetNames, ref)
+			i.targetName = v
+		case keyBuildInfo:
+			if v == "" {
+				i.buildInfo = true
+				continue
+			}
+			b, err := strconv.ParseBool(v)
+			if err != nil {
+				return nil, errors.Wrapf(err, "non-bool value specified for %s", k)
```
Comment (thaJeztah) [inline builder/builder-next/exporter/export.go:64]: We need to have a look at the error-types returned (and add a proper "invalid parameter" type); perhaps we need to consider moving some of this paring into the api side, and use `httputils.BoolValue()` (or a variant thereof, as it feels a bit _too_ permissive for an API); https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/api/server/httputils/form.go#L10-L13

There's already a bunch of these in this package, so let's do so separately.

File: builder/builder-next/exporter/writer.go [hunk 7]
```diff
@@ -20,13 +21,25 @@ import (
...
-	img := ocispec.Image{
-		Architecture: runtime.GOARCH,
-		OS:           runtime.GOOS,
+	pl := platforms.Normalize(platforms.DefaultSpec())
+
+	type image struct {
+		ocispec.Image
+
+		// Variant defines platform variant. To be added to OCI.
+		Variant string `json:"variant,omitempty"`
```
Comment (thaJeztah) [inline builder/builder-next/exporter/writer.go:30]: Looks like the vendoring is updated in this PR to a version that has the `Variant` field; so we probably don't need this custom struct? https://github.com/moby/moby/pull/43239/commits/7eb4ef66775075b9c644155ccab8ae5e08d3e51c#diff-cbddc5af26a81e622cffefb37996ea04e1b66aa3a811d8b9e9b291ffb8195d19R92-R94

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Starting to look good; left one comment about the image struct we probably no longer need, and some notes for follow-ups (no blocker for this PR).

Perhaps some of the remaining dependencies could also already be updated (klauspost compress, e.g.), but that's mostly to make the diff even easier to review 😅)

File: builder/builder-next/exporter/writer.go (hunk 7 above)
Comment (crazy-max) [reply builder/builder-next/exporter/writer.go]: Done

File: builder/builder-next/adapters/snapshot/snapshot.go [hunk 8]
```diff
@@ -242,7 +243,7 @@ func (s *snapshotter) Stat(ctx context.Context, key string) (snapshots.Info, err
 	if err := s.db.View(func(tx *bolt.Tx) error {
 		b := tx.Bucket([]byte(id))
 		if b == nil && l == nil {
-			return errors.Errorf("snapshot %s not found", id) // TODO: typed
+			return errdefs.ErrNotFound
```
Comment (thaJeztah) [inline builder/builder-next/adapters/snapshot/snapshot.go:246]: I think these can be wrapped if the additional information is useful, e.g.;

```go
return errors.Wrapf(errdefs.ErrNotFound, "snapshot %s", id)
```

That's similar to how containerd itself uses these; https://github.com/moby/moby/blob/d5d5f258dfc95c46ed1e62953a754b7cf3edecd3/vendor/github.com/containerd/containerd/container.go#L201

**Comment by crazy-max** [comment]
Seems like a flaky test on win-2022-c8d: https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-43239/45/pipeline#step-271-log-1702

Are you aware of this one @thaJeztah?

```
[2022-03-22T14:41:24.835Z] === Failed
[2022-03-22T14:41:24.835Z] === FAIL: github.com/docker/docker/integration-cli TestDockerSuite/TestRunContainerWithRmFlagCannotStartContainer (2.65s)
[2022-03-22T14:41:24.835Z]     docker_cli_run_test.go:2770: Expected not to have containers 918947e84a9f
[2022-03-22T14:41:24.835Z]         
[2022-03-22T14:41:24.835Z]     --- FAIL: TestDockerSuite/TestRunContainerWithRmFlagCannotStartContainer (2.65s)
[2022-03-22T14:41:24.835Z] 
[2022-03-22T14:41:24.835Z] === FAIL: github.com/docker/docker/integration-cli TestDockerSuite (3009.20s)
```

File: builder/builder-next/adapters/snapshot/snapshot.go (hunk 8 above)
Comment (tonistiigi) [reply builder/builder-next/adapters/snapshot/snapshot.go:246]: @crazy-max ^

**Review by tonistiigi — COMMENTED** [review]
Make sure the leftover mergeop/hardlink issue is tracked in moby with 22.x milestone so we can get this in for early testing.

**Review by tonistiigi — APPROVED** [review]

File: builder/builder-next/exporter/writer.go [hunk 9]
```diff
@@ -173,33 +175,39 @@ func normalizeLayersAndHistory(diffs []digest.Digest, history []ocispec.History,
...
-		defer p.Release(context.TODO())
+
+	metas := make([]refMetadata, len(layerChain))
+	for i, layer := range layerChain {
+		meta := &metas[i]
+
+		if description := layer.GetDescription(); description != "" {
+			meta.description = description
+		} else {
+			meta.description = "created by buildkit" // shouldn't be shown but don't fail build
```
Comment (thaJeztah) [inline builder/builder-next/exporter/writer.go:211]: Was wondering if we use this information anywhere (and if so, if this at some point should include the buildkit version)

File: builder/builder-next/builder.go [hunk 10]
```diff
@@ -319,6 +320,17 @@ func (b *Builder) Build(ctx context.Context, opt backend.BuildConfig) (*builder.
 	}
 	frontendAttrs["add-hosts"] = extraHosts
 
+	if opt.Options.ShmSize > 0 {
+		frontendAttrs["shm-size"] = strconv.FormatInt(opt.Options.ShmSize, 10)
+	}
+
+	ulimits, err := toBuildkitUlimits(opt.Options.Ulimits)
+	if err != nil {
+		return nil, err
```
Comment (thaJeztah) [inline builder/builder-next/builder.go:329]: These are also the ones we need to start looking at to return a correct error-type (InvalidParam()); per my other comment; there's a lot of those already, so let's make an effort to clean that up in some sweeps.

File: builder/builder-next/builder.go [hunk 11]
```diff
@@ -558,6 +570,18 @@ func toBuildkitExtraHosts(inp []string) (string, error) {
 	return strings.Join(hosts, ","), nil
 }
 
+// toBuildkitUlimits converts ulimits from docker type=soft:hard format to buildkit's csv format
+func toBuildkitUlimits(inp []*units.Ulimit) (string, error) {
```
Comment (thaJeztah) [inline builder/builder-next/builder.go:574]: oh... hmm... so actually looks like this never return an error? ok to fix later (somewhat surprised there's no longer complaining)

File: builder/builder-next/exporter/writer.go [hunk 12]
```diff
@@ -80,13 +81,23 @@ func patchImageConfig(dt []byte, dps []digest.Digest, history []ocispec.History,
 	}
 
 	if cache != nil {
-		dt, err = json.Marshal(cache)
+		dt, err := json.Marshal(cache)
 		if err != nil {
-			return nil, errors.Wrap(err, "failed to marshal cache")
```
Comment (thaJeztah) [inline builder/builder-next/exporter/writer.go:83]: Somewhat curious why you removed the wrapping here (seems like the additional context could be useful); not a blocker, but add it to the list of things to look at

**Review by thaJeztah — APPROVED** [review]
LGTM - left some notes about things we should look into in follow-ups

**Comment by thaJeztah** [comment]
> Make sure the leftover mergeop/hardlink issue is tracked in moby with 22.x milestone so we can get this in for early testing.

tracked in https://github.com/moby/moby/issues/43415

**Comment by thaJeztah** [comment]
Let's get this one in!

**Pull request merged**